### PR TITLE
net: http: define http service and resource iterable sections

### PR DESCRIPTION
**Pull request by cfriedt**
Provide a means of declaring zero or more HTTP services, each with zero or more static HTTP resources.

Static HTTP resources are those which have fixed paths[1] that are known prior to system initialization. Some examples of static http resources would be    
* a forwarder from '/' to '/index.html'
* a REST endpoint with fixed path '/api/foo' and detail pointing at some implementation-specific function
* a Javascript file in string form with fixed path '/js/util.js'
* a 'construction' image with path '/res/work.png'
* a gzip-compressed 'Hello' HTML file at '/hello.html'
    
Without loss of generality in how static HTTP resources are organized or served by any given HTTP server, we can describe what static resources exist on a system in a common way that
* does not require any optional facilities (e.g. filesystem) and
* relies exclusively on addressable memory.

Additionally, for the purposes of simply allowing others to implement custom HTTP servers in a consistent way, or benchmarking implementations, or having a consistent testsuite to use across multiple implementations, it is helpful to have a common method to declare HTTP services and static resources for Zephyr.

[1] https://en.wikipedia.org/wiki/URL

### CODE REVIEW

**Comment by cfriedt** [comment]
Blocked on https://github.com/zephyrproject-rtos/zephyr/pull/55429

**Comment by cfriedt** [comment]
Also, just to clarify, yes, this does allow multiple HTTP servers to operate simultaneously (or a single server, serving multiple ports).

**Comment by cfriedt** [comment]
* removed `BUILD_ASSERT()` from `service.h` because llvm was throwing errors of the form 
```
error: static_assert expression is not an integral constant expression
```

**Comment by cfriedt** [comment]
I woudl say that possibly the most useful result of this change is that it allows us to test all HTTP server implementations using the same code.

**Comment by cfriedt** [comment]
This is important to get in well before GSoC contributors finish preparing their applications / sample code.

**Comment by cfriedt** [comment]
* implement `STRUCT_SECTION_FOREACH()` in terms of `STRUCT_SECTION_FOREACH_ALTERNATE()`

File: tests/net/lib/http_server/common/src/main.c
Comment (jukkar) [inline tests/net/lib/http_server/common/src/main.c:93]: Perhaps we could follow the style guide here, so { to previous line. Same comment to rest of the file

**Review by jukkar — COMMENTED** [review]
Looks cool, nice work :+1:

File: tests/net/lib/http_server/common/src/main.c
Comment (cfriedt) [reply tests/net/lib/http_server/common/src/main.c:93]: Something is broken in `zephyr/.clang-format`.

To fix, I have to specifically disable it and reformat these types of macros manually, which is a bit of an annoyance.

Will do.

Comment (cfriedt) [reply tests/net/lib/http_server/common/src/main.c]: Fixed!

File: include/zephyr/net/http/service.h
Comment (rlubos) [inline include/zephyr/net/http/service.h:47]: Any particular reason for `port` to be a pointer? If it's not intended to be NULL in any case, perhaps just store the actual port value in the structure?

File: include/zephyr/net/http/service.h
Comment (rlubos) [inline include/zephyr/net/http/service.h:97]: And if it's not writeable? Sounds like a potential source od problems (and ambiguity in behaviour).

File: subsys/net/lib/http/Kconfig
Comment (rlubos) [inline subsys/net/lib/http/Kconfig:37]: Maybe add some note it's WiP here?

File: include/zephyr/net/http/service.h
Comment (cfriedt) [reply include/zephyr/net/http/service.h:47]: Because it could be ephemeral. It's similar to the mDNS-SD port

File: include/zephyr/net/http/service.h
Comment (cfriedt) [reply include/zephyr/net/http/service.h:97]: Then it would read as whatever value it was initialized with.

Comment (cfriedt) [reply include/zephyr/net/http/service.h:97]: Should be possible for an implementation to check if a memory region is RO, IIRC.

**Review by rlubos — DISMISSED** [review]

**Comment by cfriedt** [comment]
@rlubos @jukkar - taken out of draft. Please pre-approve if it looks ok.

**Review by jukkar — APPROVED** [review]
LGTM

**Review by rlubos — APPROVED** [review]

**Pull request merged**
